### ISSUE TITLE
/dev/dri/card0 creation is significantly delayed on boot 

### ISSUE THREAD

**Issue by birdie-github** [systemd/systemd#25408]
### systemd version the issue has been seen with

systemd-251.8-586.fc37.x86_64

### Used distribution

Fedora 37

### Linux kernel version used

6.0.8-300.fc37.x86_64

### CPU architectures issue was seen on

x86_64

### Component

systemd-udevd

### Expected behaviour you didn't see

Since I've upgraded from F36 to F37 Xorg fails to start for at least 10 seconds on boot because the mentioned file is missing.

This is a regression.

Nothing on my system has changed configuration-wise.

Check 

https://bugzilla.redhat.com/show_bug.cgi?id=2143317
https://bugzilla.redhat.com/show_bug.cgi?id=2143248

For more info.

This is reproducible with the kernel and xorg-x11-server-Xorg packages from Fedora 36 (temporarily installed them), so they are not the culprit.

I'd be glad to try systemd-udev from Fedora 36 but I don't want to break my build:

```
# rpm -Uvh systemd-udev-250.8-1.fc36.x86_64.rpm 
error: Failed dependencies:
	libsystemd-shared-250.so()(64bit) is needed by systemd-udev-250.8-1.fc36.x86_64
	libsystemd-shared-250.so(SD_SHARED)(64bit) is needed by systemd-udev-250.8-1.fc36.x86_64
	systemd(x86-64) = 250.8-1.fc36 is needed by systemd-udev-250.8-1.fc36.x86_64
```

**Comment by yuwata** [comment]
This is a kind of duplicate of #24026. That is, X is now started too early. See the logs you uploaded to RHBZ, in the failed case, X starts after 10sec, but on success, it starts after 16sec.

I am not familiar about unit dependencies of X or early starting stage of X, but it should wait for a dri device is created.
Which service unit starts X? And, what dependencies set for the unit?

**Comment by birdie-github** [comment]
I've been using LXDM for the past ten years or so.

```
[Unit]
Description=LXDM (Lightweight X11 Display Manager)
#Documentation=man:lxdm(8)
Conflicts=getty@tty1.service
After=systemd-user-sessions.service getty@tty1.service plymouth-quit.service livesys-late.service
#Conflicts=plymouth-quit.service

[Service]
ExecStart=/usr/sbin/lxdm
Restart=always
IgnoreSIGPIPE=no
#BusName=org.freedesktop.lxdm

[Install]
Alias=display-manager.service
```
It's exactly the same as in Fedora 36 which was free from this bug.

**Comment by birdie-github** [comment]
> This is a kind of duplicate of #24026. That is, X is now started too early. See the logs you uploaded to RHBZ, in the failed case, X starts after 10sec, but on success, it starts after 16sec.
> 
> I am not familiar about unit dependencies of X or early starting stage of X, but it should wait for a dri device is created. Which service unit starts X? And, what dependencies set for the unit?

I can't believe X.org can possibly start "too early", it's been the same over the past 7 years or so, X.org starts literally 2-3 seconds after entering LUKS2 password, that's what I expect from an SATA SSD disk (not as fast as NVMe but still a lot faster than any HDD especially when it comes to seeking which is near instant).

**Comment by yuwata** [comment]
We recently changed the order of device initialization: 2336bde96420475ccb054326f27290fa0228f27d, 12bdeb58a63d790f5f83fc45819f61029c0d89bd, 403ca5b8b4e5092639d350f84994b4cfb357e1aa.
I guess that causes drm devices initialized slightly slower than previous, and other devices faster, and X started slightly earlier.

**Comment by keszybz** [comment]
I'll copy my answer from https://bugzilla.redhat.com/show_bug.cgi?id=2143248:

This isessentially a race condition. Because of some changes in which devices are initialized first, /dev/dri/card* is apparently initialized more slowly than in the past. It's not entirely clear why that is, but the kernel simply needs some time to handle devices, and they don't always show up "immediately". The race condition was always there, we just were lucky not to see it.

There isn't much systemd can do. Please either make the Xorg server handle this case correctly or add some ordering to make the unit wait until the device appears.

**Comment by MichaelJTokarev** [comment]
What is the syncronisation point for Xorg in this case, ie, how this race can be soled?  I understand quite well this is not a systemd problem, but what is the solution to use outside of systemd?  I'm starting a GUI session from a .service file, using startx or sway/weston, whatever, — what should I put in After= or Wants= for this to be sure it is started after /dev/dri/card0 is here?

I'm not sure this can be just closed without at least one suggested solution.. :)

**Comment by mbiebl** [comment]
As mentioned on IRC: tag the device via `TAG+="systemd"` in a udev rule. This will make it show up as a .device unit which you can then use as `After=<foo>.device` in your .service unit.

**Comment by birdie-github** [comment]
This bug is weird. I swear it disappeared for a while with Linux 6.0, now I've upgraded to Linux 6.1.5-200.fc37.x86_64 and it's back again:

```
[    11.927] (II) VESA: driver for VESA chipsets: vesa
[    11.928] xf86EnableIOPorts: failed to set IOPL for I/O (Operation not permitted)
[    11.928] (EE) open /dev/dri/card0: No such file or directory
[    11.928] (WW) Falling back to old probe method for modesetting
[    11.928] (EE) open /dev/dri/card0: No such file or directory
[    11.928] vesa: Ignoring device with a bound kernel driver
[    11.928] (EE) Screen 0 deleted because of no matching config section.
[    11.928] (II) UnloadModule: "modesetting"
[    11.928] (EE) Screen 0 deleted because of no matching config section.
[    11.928] (II) UnloadModule: "vesa"
[    11.928] (EE) Device(s) detected, but none match those in the config file.
[    11.928] (EE)
Fatal server error:
[    11.928] (EE) no screens found(EE)
[    11.928] (EE)
```

**Comment by birdie-github** [comment]
I still do not understand why a creation of a single file representing my HW takes so much time. It must be instant for all I know as long as `i915.ko` gets loaded which happens **before** the system actually starts booting because the module is in the initrd.

With Fedora 36 I never had this bug, lxdm took maybe 2.5 seconds to show up after the kernel finished booting.

With Fedora 37 I need to wait for around 12 seconds because something is delayed.

Why? Who's responsible? What's going on?

Systemd is largely developed by RedHat employees who also largely overlook Fedora's development. Please do not tell me to use Fedora's bugzilla. I'm freaking tired of posting bugs over there - they are often never addressed, they often receive zero comments.

**Comment by MichaelJTokarev** [comment]
> This bug is weird. I swear it disappeared for a while with Linux 6.0, now I've upgraded to Linux 6.1.5-200.fc37.x86_64 and it's back again:

It haven't disappeared. It is random, because there's no proper dependencies, and the failure depends on the number of more or less random factors, — timing. Some combination of software will trigger it some wont. Sometimes a tiny delay before starting                                  X will be enough to work around this issue, but it might break again on the next change.

**Comment by birdie-github** [comment]
> It haven't disappeared. It is random, because there's no proper dependencies, and the failure depends on the number of more or less random factors, — timing. Some combination of software will trigger it some wont. Sometimes a tiny delay before starting X will be enough to work around this issue, but it might break again on the next change.

What should I edit to make this file go up as soon as the i915 module gets loaded?

```
# lsmod | grep i915
i915                 3436544  33
drm_buddy              20480  1 i915
ttm                    94208  1 i915
drm_display_helper    208896  1 i915
cec                    81920  2 drm_display_helper,i915
video                  65536  1 i915
```

**Comment by MichaelJTokarev** [comment]
> As mentioned on IRC: tag the device via `TAG+="systemd"` in a udev rule. This will make it show up as a .device unit which you can then use as `After=<foo>.device` in your .service unit.

/etc/udev/rules.d/dev-dri-card0.rules:
 ```
ENV{DEVNAME}=="/dev/dri/card0", TAG+="systemd"
```
And in the unit file:
```
Wants=dev-dri-card0.device
After=dev-dri-card0.device
```
This is still ugly but works in simple case.  This apparently needs to be changed into seat dependencies instead, ie, require for the given seat (seat0 by default) before starting this service.

**Comment by birdie-github** [comment]
Sadly this is still unresolved as of Fedora 38, `systemd-253.7-1.fc38.x86_64`.

And now I have a different laptop with `amdgpu` instead of `i915`.
